Create auth client

SiQos opened the pull request:
- [x]  Create better-auth like hono docs
- [x] Create log out button and remove email's input
- [x] Create change theme without lib
- [x] Fix directory
- [x] Delete useless package
- [x] Add our config to .gitignore
- [x] Fix theme context
- [x] Create prettier config

SiQos committed:
feat: create auth function with better-auth lib and create header

SiQos committed:
feat: create auth, create header component, add dark/light theme

SiQos committed:
fix: delete useless function for theme
docs: add prettier config
feat: create component for handle errors, add correct session check func

SiQos committed:
feat: create change theme func, add error UI handling for auth

SiQos committed:
fix: fix incorrect get session

SiQos committed:
fix: fix initial view theme component

SiQos committed:
fix: fix incorrect get theme from storage

SiQos committed:
refactor: divide the components into different shared folders

SiQos committed:
docs: delete todo file

SiQos committed:
docs: delete .vscode file

alexatsu commented on .prettierrc:
Ничего не написал сюда.
Добавь:
{
  "tabWidth": 2,
  "singleQuote": true,
  "printWidth": 120
}

alexatsu commented on README.md:
````diff
@@ -0,0 +1,10 @@
+Create a `.env` file (if it doesn’t exist) and add the following variable:
+
+```env
+NEXT_PUBLIC_API_URL="https://your-api-url.com"
````
Добавь еще это:
### setup dependencies

```pnpm i```

### start the project

```pnpm dev```

alexatsu commented on app/(private)/_shared/components/error.tsx:
```diff
@@ -0,0 +1,27 @@
+'use client';
```
Давай это изменим, чтобы юзер сам выбрал куда ему надо. На логин или в поддержку (дискорд). Не перенаправляем сами

SiQos replied:
Done

alexatsu commented on app/(private)/_shared/components/header.tsx:
```diff
@@ -0,0 +1,39 @@
+            onClick={() =>
+              signOut({
+                fetchOptions: {
+                  onSuccess: () => redirect("/login"),
```
Это клиентский компонент. Меняем redirect на useRouter

SiQos replied:
Done

alexatsu commented on app/(private)/_shared/components/header.tsx:
```diff
@@ -0,0 +1,39 @@
+            Sign Out
+          </Button>
+          <Avatar>
+            <AvatarImage src={user.image || ""} />
```
Если аватара нет у пользователя, тогда будет лучше добавить placeholder 
<img width="363" height="138" alt="{75BC33EF-1228-4D7D-90F0-CEE60671910F}" src="https://github.com/user-attachments/assets/67452e51-74b9-4bc1-920b-4aae63b02a75" />

SiQos replied:
Done

alexatsu commented on app/(private)/_shared/components/header.tsx:
```diff
@@ -0,0 +1,39 @@
+        </>
+      ) : (
+        <>
+          <Button onClick={() => redirect("/login")}>Sign in</Button>
```
useRouter

alexatsu commented on app/(private)/_shared/components/header.tsx:
```diff
@@ -0,0 +1,39 @@
+      ) : (
+        <>
+          <Button onClick={() => redirect("/login")}>Sign in</Button>
+        </>
```
фрагмент убираем

alexatsu commented on app/(private)/_shared/components/theme.tsx:
```diff
@@ -0,0 +1,57 @@
+export type Theme = keyof typeof ThemeType;
+
+const getThemeFromLS = () =>
+  localStorage.getItem(localStorageThemeKey) as ThemeType | null;
```
as Theme | null. Одна же тема в ls хранится

alexatsu commented on app/(private)/_shared/components/theme.tsx:
```diff
@@ -0,0 +1,57 @@
+  const [theme, setTheme] = useState<Theme>();
+
+  const changeTheme = () => {
+    if (theme) {
```
theme всегда есть, если судить по типам. Убираем условие

alexatsu commented on app/(private)/_shared/components/theme.tsx:
```diff
@@ -0,0 +1,57 @@
+
+  const changeTheme = () => {
+    if (theme) {
+      document.documentElement.setAttribute("data-theme", theme);
```
Магическая строка => в переменную

alexatsu commented on app/(private)/_shared/components/theme.tsx:
```diff
@@ -0,0 +1,57 @@
+  useEffect(() => {
+    const theme = getThemeFromLS();
+    if (theme) {
+      setThemeToLS(theme);
```
Точно надо?

alexatsu commented on app/(private)/_shared/components/theme.tsx:
```diff
@@ -0,0 +1,57 @@
+      setThemeToLS(theme);
+      setTheme(theme);
+    }
+  }, []);
```
Пробел между функциями, или используй formatter

alexatsu commented on app/(private)/_shared/lib/getUser.ts:
```diff
@@ -0,0 +1,21 @@
+import { APIError } from "better-auth";
+import { getSession } from "@/lib/auth";
+
+export async function getUser() {
```
Получаем юзера в одном месте => положи в layout все тогда

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -0,0 +1,18 @@
+import { Header } from "@/app/(private)/_shared/components/header";
```
Как же хочется пробелов в функции, хочется дышать

alexatsu commented on app/(public)/_shared/components/error.tsx:
```diff
@@ -0,0 +1,27 @@
+'use client';
```
Этот файл убрать и добавить global error. В nextjs документации ответ.
Если нужны error кастомный, тогда делаем. Пока нет нужды

alexatsu commented on app/(public)/_shared/components/error.tsx:
все паблики удаляем

alexatsu commented on app/(public)/_shared/components/header.tsx:
все паблики удаляем

alexatsu commented on app/(public)/_shared/components/theme.tsx:
все паблики удаляем

alexatsu commented on app/(public)/_shared/lib/getUser.ts:
все паблики удаляем

SiQos replied:
Done

alexatsu commented on app/(public)/login/_shared/components/LoginForm.tsx:
все паблики удаляем

SiQos replied:
Done

alexatsu commented on app/(public)/login/page.tsx:
все паблики удаляем

SiQos replied:
Done

alexatsu commented on app/page.tsx:
```diff
@@ -0,0 +1,25 @@
+
+  if (isPending) return <Spinner className="absolute top-2/4 left-2/4" />;
+  if (error) return <Error error={error.message ?? "Unknown error"} />;
+  if (!session) {
```
Обсудили в звонке. 4 состояния приложения (2 сессии есть/нет) (2 сервера доступен/нет)

SiQos replied:
Done

alexatsu commented on components.json:
Переместить все что связано с shadcn в папку root/app/_shared

SiQos replied:
Done

SiQos committed:
fix: add error component for handling fetch error
refactor: split auth logic into other function
config: add some config to prettier, add path resolver to eslint

SiQos committed:
fix: add correct get and set theme to storage and fix wrong view theme's icons

SiQos committed:
Merge branch 'task#2' of https://github.com/MidnightNebula/nebula-admin into task#2

SiQos committed:
feat: add default avatar if it isn't

SiQos committed:
Revert "feat: add default avatar if it isn't"

SiQos committed:
"feat: add default avatar if it isn't"

SiQos committed:
"feat: add default avatar if it isn't"

alexatsu commented on app/(features)/dashboard/layout.tsx:
```diff
@@ -0,0 +1,14 @@
+import { redirect } from 'next/navigation';
+
+export default async function Layout({ children }: { children: React.ReactNode }) {
+  const user = await getUser();
```
пробел

SiQos replied:
Done

alexatsu commented on app/(features)/dashboard/layout.tsx:
```diff
@@ -0,0 +1,14 @@
+
+export default async function Layout({ children }: { children: React.ReactNode }) {
+  const user = await getUser();
+  if (!user) redirect('/login');
```
пробел

SiQos replied:
Done

alexatsu commented on app/(features)/login/page.tsx:
```diff
@@ -0,0 +1,27 @@
+      <div className="flex flex-col gap-4 p-6 md:p-10">
+        <div className="flex flex-1 items-center justify-center">
+          <div className="w-full max-w-xs">
+            <LoginForm onSubmit={() => discordAuth({ setError, setIsPending })} />
```
discordAuth положи сюда

SiQos replied:
Done

alexatsu commented on app/(features)/dashboard/layout.tsx:
```diff
@@ -0,0 +1,14 @@
+  if (!user) redirect('/login');
+  return (
+    <div>
+      <Header user={user} />
```
Почему Header в двух местах? Кладем в global layout один раз и все

SiQos replied:
Done

alexatsu commented on app/_shared/components/LoadingSpinner.tsx:
Этот компонент нужен? Обертка над оберткой

SiQos replied:
Done

alexatsu commented:
Все компоненты переименовать в PascalCase

alexatsu commented on app/_shared/components/Theme.tsx:
```diff
@@ -0,0 +1,50 @@
+    } else {
+      setTheme(ThemeType.light);
+    }
+  }, []);
```
пробел

SiQos replied:
Done

alexatsu commented on app/_shared/components/header.tsx:
```diff
@@ -0,0 +1,39 @@
+import { Avatar, AvatarImage } from '@/app/_shared/shadcn/components/avatar';
+import { Theme } from '@/app/_shared/components/theme';
+
+export function Header({ user }: { user: User }) {
```
header используется два раза, проверка на юзера получается тоже. Оставить одну проверку, и только в layout

SiQos replied:
Done

alexatsu replied:
Так а почему два раза header используется? Оставить один раз.

SiQos committed:
feat: add positive UI when:
1) Session is not found, server is working - redirect to login page
2) Session is found, server does not working - positive error
3) Session is not found, server does not work - redirect to login page

SiQos committed:
fix: fix flicking icon

SiQos committed:
fix: fix no network handling function

alexatsu commented:
Проверить компоненты, которые названы с нижнего регистра. header => Header, component => Component, etc
<img width="1006" height="420" alt="{6419F571-4672-4E99-8EDA-68A29DEF42A8}" src="https://github.com/user-attachments/assets/01472d95-f9d2-4f65-b559-7a1c93e17eb4" />

alexatsu commented:
ci/.env.example где?
<img width="1147" height="856" alt="{A04BE6AB-F026-43D7-B3D7-EA9F5C1EAE61}" src="https://github.com/user-attachments/assets/dabc2e32-8b09-42c6-9fb4-c701c83a2912" />

<img width="890" height="130" alt="{8489999C-90A4-43FF-BC5F-A45CA5AD0808}" src="https://github.com/user-attachments/assets/699fe746-bf3b-4a3b-a7df-af616a0b1b63" />

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -0,0 +1,17 @@
+  return (
+    <>
+      <Header user={user} />
+      {isAuthorized && error ? <ErrorTooltip /> : children}
```
isAuthorized не надо

alexatsu commented on app/(public)/login/_shared/components/LoginForm.tsx:
```diff
@@ -0,0 +1,69 @@
+            type="submit"
+            disabled={isPending}
+          >
+            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor">
```
svg положи в компонент и потом вставить сюда.

SiQos replied:
Done

alexatsu commented on app/_shared/components/ErrorTooltip.tsx:
.jsx => .tsx

SiQos replied:
Done

alexatsu commented on app/_shared/components/Header.tsx:
Этот файл используется два раза, а можем один раз?
Также header надо добавлять в layout а не в страницу. Мы не будем добавлять его везде.
<img width="432" height="296" alt="{E6497818-9C10-4BF1-BC85-EE821EDE27CE}" src="https://github.com/user-attachments/assets/f6d2e405-e4a6-4900-affa-72f03480d849" />

SiQos replied:
Done

alexatsu commented on app/_shared/lib/auth.ts:
```diff
@@ -0,0 +1,11 @@
+import { createAuthClient } from 'better-auth/react';
+
+export const authClient = createAuthClient({
+  baseURL: process.env.NEXT_PUBLIC_API_URL,
```
а в ci/.env.example его нет, пример NEXT_PUBLIC_API_URL="http://localhost:4000"

alexatsu commented on app/_shared/lib/getUser.ts:
```diff
@@ -0,0 +1,67 @@
+  };
+};
+
+async function getAuthContext() {
```
Что это? Удаляем, если не используем

alexatsu commented on app/_shared/lib/getUser.ts:
```diff
@@ -0,0 +1,67 @@
+
+    return { user, error, isAuthorized };
+  } catch (err) {
+    return { user: null, error: getError(500, 'NETWORK_NOT_FOUND'), isAuthorized };
```
Если проверяешь доступна ли сеть, тогда добавь функционал для сервера. Если ты проверяешь, если доступен сервер или нет, тогда ошибку поставь соответствующую.

alexatsu commented on app/_shared/lib/getUser.ts:
```diff
@@ -0,0 +1,67 @@
+  NETWORK_NOT_FOUND: 'Network is not found',
+};
+
+type ReturnDataSession = {
```
убрать просил, ты переписываешь тип из библиотеки

alexatsu commented on app/_shared/lib/getUser.ts:
Замени на:
```
import { authClient, getSession } from '@/_shared/lib/auth';
import { cookies, headers } from 'next/headers';

const SESSION_TOKEN_NAME = 'better-auth.session_token' as const;

const ERRORS = {
  ...authClient.$ERROR_CODES,
  NETWORK_NOT_FOUND: 'NETWORK_NOT_FOUND',
} as const;

type ErrorCode = keyof typeof ERRORS;

export async function getUser() {
  const { headers } = await getAuthContext();

  try {
    const { data: session, error } = await getSession({
      fetchOptions: {
        credentials: 'include',
        headers,
      },
    });

    const user = session?.user ?? null;
    const isAuthorized = !!session;

    return {
      user,
      error,
      isAuthorized,
    };
  } catch {
    return {
      user: null,
      error: getError(500, ERRORS.NETWORK_NOT_FOUND),
      isAuthorized: false,
    };
  }
}

async function getAuthContext() {
  const cookieStore = await cookies();

  return {
    headers: await headers(),
    hasSessionCookie: cookieStore.has(SESSION_TOKEN_NAME),
  };
}

const getError = (status: number, statusCode: ErrorCode) => {
  return {
    status,
    statusText: statusCode,
    message: ERRORS[statusCode],
  };
};
```

alexatsu commented on app/global-error.tsx:
Проверь, это global error или нет.

alexatsu commented on app/page.tsx:
```diff
@@ -0,0 +1,18 @@
+  return (
+    <>
+      <Header user={user} />
+      {error?.status === 500 && isAuthorized ? <ErrorTooltip /> : <div>Hello</div>}
```
Тут проверка отличается от версии с сессии. Почему?
&& Authorized убрать

alexatsu commented on app/page.tsx:
```diff
@@ -0,0 +1,18 @@
+
+export default async function Home() {
+  const { user, error, isAuthorized } = await getUser();
+  console.log('user 2:', user);
```
Убрать логи.

alexatsu commented on todo.md:
Зачем это в репо?

SiQos committed:
refactor: remove two headers and replace global root into private folder

SiQos committed:
refactor: change ErrorTooltip component from jsx to tsx

SiQos committed:
refactor: change svg icon to DiscordIcon component

SiQos committed:
fix: delete getUser func

SiQos committed:
fix: add correct session handler

SiQos committed:
refactor: delete md file

SiQos committed:
fix: fix global error component

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -0,0 +1,68 @@
+      <head />
+      <body className={`dark:bg-gray-900 ${geistSans.variable} ${geistMono.variable} antialiased`}>
+        <script
+          dangerouslySetInnerHTML={{
```
Этот код повторяется в проекте. Используется только в RootLayout, и этот RootLayout может быть только один.

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -0,0 +1,68 @@
+
+export default async function RootLayout({ children }: { children: React.ReactNode }) {
+  let session = null;
+  let e: any;
```
А где типы?

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -0,0 +1,68 @@
+  title: 'Create Next App',
+  description: 'Generated by create next app',
+};
+const SESSION_TOKEN_NAME = 'better-auth.session_token' as const;
```
в _shared/consts или _shared/config

alexatsu commented on app/(public)/login/_shared/components/LoginForm.tsx:
```diff
@@ -0,0 +1,68 @@
+    <form
+      className={cn('flex flex-col gap-6', className)}
+      {...props}
+      onSubmit={async (e) => {
```
логика в разметке. в функцию

alexatsu commented on app/_shared/components/header.tsx:
```diff
@@ -0,0 +1,50 @@
+import { useRouter } from 'next/navigation';
+import { useEffect, useState } from 'react';
+
+export function Header({ user = null }: { user: User | null | undefined }) {
```
const { data, error } = await getSession
data возвращает data | null. Откуда undefined?
Типы подстраиваем под библиотеку.

alexatsu commented on app/_shared/components/header.tsx:
```diff
@@ -0,0 +1,50 @@
+import { useEffect, useState } from 'react';
+
+export function Header({ user = null }: { user: User | null | undefined }) {
+  const [isMounted, setIsMounted] = useState(false);
```
Создай хук useMounted 
Пример кода (есть лучше):
```ts
export function useMounted() {
  const [mounted, setMounted] = useState(false);
  useEffect(() => setMounted(true), []);
  return mounted;
}
```

alexatsu commented on app/_shared/components/header.tsx:
```diff
@@ -0,0 +1,50 @@
+        {user && (
+          <>
+            <Button
+              onClick={() =>
```
функционал из разметки убираем

alexatsu commented on tsconfig.json:
Проверить компоненты, которые названы с нижнего регистра. header => Header, component => Component, etc
<img width="1006" height="420" alt="image" src="https://github.com/user-attachments/assets/22663ca7-cfcb-4594-9773-39cdfcbca2d3" />

alexatsu commented on tsconfig.json:
ci/.env.example где?
<img width="1147" height="856" alt="image" src="https://github.com/user-attachments/assets/a0e2bbb2-4a7e-4911-bda1-53dab2e39e66" />

SiQos committed:
refactor: fix types

SiQos committed:
feat: create useMounted hook

SiQos committed:
refactor: replace logOut logic

SiQos committed:
docs: delete env.example from gitignore

SiQos committed:
chore: rename components

SiQos committed:
chore: rename component

SiQos committed:
feat: create CenterContainer component

SiQos committed:
Merge branch 'task#2' of https://github.com/MidnightNebula/nebula-admin into task#2

SiQos committed:
refactor: replace setTheme from Html to lib function

alexatsu commented on app/_shared/config/seo.ts:
А чего это отдельно, в layout не надо?

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -1,23 +1,19 @@
-  let session = null;
-  let e: any;
+  let session: Session | null = null;
+  let e: string | undefined;
```
string | null

alexatsu commented on app/_shared/components/Header.tsx:
```diff
@@ -1,21 +1,24 @@
 
-export function Header({ user = null }: { user: User | null | undefined }) {
-  const [isMounted, setIsMounted] = useState(false);
+export function Header({ user = null }: { user: User | null }) {
```
user есть всегда здесь, мы чекаем это в layout, убрать null

alexatsu commented on app/_shared/lib/setTheme.ts:
```diff
@@ -0,0 +1,14 @@
+function setTheme() {
```
try catch здесь не нужен

alexatsu commented on app/(private)/layout.tsx:
```diff
@@ -0,0 +1,64 @@
+        <script
+          dangerouslySetInnerHTML={{
+            __html: `
+            (function() {
```
сюда тоже переиспользовать надо
